### ISSUE TITLE
Feature Request: Tuples

### ISSUE THREAD

**Issue by Namek** [mint-lang/mint#193]
It would be nice to have tuples such as pairs, triples and so on. I am not sure if that should be syntax thing or just a module (however then, with #191 done).

A dependency to this may be having **hole types in records** (mentioned in #130) or just a built-in module similar to `Array`, `Map` or `Set`.

## Why do we need a Tuple?

Sometimes we want to return a more complicated result in code of general usage, like packages.

For example, I have written myself a function like this in Elm:

```elm
{-| Find an element and return it along with the rest of the input items.
-}
findAndGetRest : (a -> Bool) -> List a -> ( Maybe a, List a )
findAndGetRest predicate list =
    let
        iterate checked unchecked =
            case unchecked of
                [] ->
                    ( Nothing, list )

                first :: rest ->
                    if predicate first then
                        ( Just first, List.append (List.reverse checked) rest )

                    else
                        iterate (first :: checked) rest
    in
    iterate [] list
```

which, I believe, is a good candidate for some publicly shared package.

I could have more functions returning tuples and for each one I would may have to create possibly a separate record with some unnecessary names.

## Current solution

Currently it can be done using a phantom enum/type.

```
enum FirstRest(a) {
  A(Maybe(a), Array(a))
}

module Util.Collection {
  /* Find an element and return it along with the rest of the input items. */
  fun findAndGetRest (predicate : Function(a, Bool), items : Array(a)) : FirstRest(a) {
    iterate([], items)
  } where {
    iterate =
      (checked : Array(a), unchecked : Array(a)) : FirstRest {
        try {
          case (unchecked[0]) {
            Maybe::Nothing => FirstRest::A(Maybe::Nothing, [])

            Maybe::Just first =>
              try {
                rest =
                  Array.slice(1, Array.size(unchecked))

                if (predicate(first)) {
                  FirstRest::A(Maybe::Just(first), Array.push(checked, rest))
                } else {
                  iterate(Array.push(first, checked), rest)
                }
              }
          }
        }
      }
  }
}
```

## How?

Tuple syntax is cool but it's not that simple to implement. Very easy solution would be to have a `Tuple` module, supporting record constructors. However, it is so easy to write one byself in a project, that it's no need to put it into API until we'll be sure what we need. In the other hand, writing `Tuple.pair(a, b)` each time instead of `{a, b}` or `(a, b)` (or w/e syntax) is not handsome.

**Comment by gdotdesign** [comment]
I aggre that tuples would be a nice addition to the language and I think basic tuples can be implemented fairly easily since:
- their type signature is calculable from their items
- they have fixed size
- their type would be `Tuple(a,b,c,d,e)`
- the array access operator could be used to access it's values `[]` without needing a `Maybe` (if the index is not an expression)

The more problematic parts is the deconstruction, which I'm not sure how to implement yet.

Syntax wise I'm not sure what would be the best:
- `{a, b}` is conflicting with records
- `(a, b)` is conflicting with parenthesized expressions but can be resolved

**Comment by Namek** [comment]
`<a, b, c>` ?

**Comment by Namek** [comment]
Another example in Elm:

```Elm
( to, from, pos ) =
    if (abs <| neighbourX - islandX) > (abs <| neighbourY - islandY) then
        ( neighbourX, islandX, rescaledX )

    else
        ( neighbourY, islandY, rescaledY )

distancePercent =
    pos |> rescale from to 0.0 1.0 |> Basics.min 1.0
```

This calculates 3 values based on one condition. Currently in Mint, if I wanted to evaluate the condition only once (because of performance, readability and maintainability) I would have to cache it into variable and use that 3 times to define `to`, `from` and `pos` separately:

```
try {
  cond = Math.abs(neighbourX - islandX) > Math.abs(neighbourY - islandY)

  to = if (cond) { neighbourX } else { neighbourY }
  from = if (cond) { islandX } else { islandY }
  pos = if (cond) { rescaledX } else { rescaledY }

  distancePercent =
    pos |> rescale(from, to, 0, 1) |> Math.min(1)
}
```

which of course grows largely with the default formatter:

```try {
  cond =
    Math.abs(neighbourX - islandX) > Math.abs(neighbourY - islandY)

  to =
    if (cond) {
      neighbourX
    } else {
      neighbourY
    }

  from =
    if (cond) {
      islandX
    } else {
      islandY
    }

  pos =
    if (cond) {
      rescaledX
    } else {
      rescaledY
    }

  distancePercent =
    pos
    |> rescale(from, to, 0, 1)
    |> Math.min(1)
}
```

**Comment by Namek** [comment]
`|a, b, c|` ?

**Comment by Namek** [comment]
Here's an example how dirty I dealt with it to have general tuples using phantom enums:

https://sandbox.mint-lang.com/sandboxes/znHxzihX20KP7Q

```
enum Pair(a, b) {
  A(a, b)
}
enum Triple(a, b, c) {
  A(a, b, c)
}

component Main {
	fun testTriple(triple : Triple(Number, Number, Number)) : Number {
  	case (triple) {
      Triple::A a b c => a + b + c
    }
  }
  
  fun render : Html {
  	<div>
  	  <{ testTriple(Triple::A(1, 2, 3)) |> Number.toString() }>
    </div>
  }
}
```

**Comment by gdotdesign** [comment]
Thanks for the example, I've been started working on this in the `tuples` branch.
